rfcs/20231005-hybrid-quantization.md

```diff
@@ -0,0 +1,158 @@
+# RFC: Hybrid quantized ops for weight-only quantization
+
+Status: Review<br/>
+Initial version: 10/05/2023<br/>
+Last updated: 10/17/2023<br/>
+
+## Version log
+
+* 10/05/2023: Initial version.
+* 10/17/2023: Minor fixes and add proposed operation semantics. 
+
+## Introduction
+
+This RFC proposes to extend the convolution and dot_general ops to allow
+differing element types for their operands and define their semantics to
+represent weight-only quantization. This work is related to the more general
+topic of [mixed precision](https://github.com/openxla/stablehlo/issues/369), but
+does not aim to address it in its entirety.
+
+Weight-only quantization is a quantization scheme which is often used to reduce
+the size of models that are memory-bound. A weight-only quantized graph accepts
+input in float and weight in quantized type, dequantizes the weight and performs
+float computation. Even though weight-only quantization can already be expressed
+using a combination of uniform_dequantize op and float-only op, we propose to
+extend the semantics of some float-only ops as a representation of weight-only
+quantization because of the following reasons.
+
+* It aligns well with the existing behavior of the corresponding HLO ops which
+allow operand types to differ and upcast operands to the highest-precision type
+in such a case.
+* Being able to express weight-only natively in StableHLO, it does not rely on
+pattern matching by backend to avoid patterns that might result in bad
+performance.
+* Constant and dequantize can unintentionally be folded in downstreams and
+hybrid op removes concern on constant folding. We can also consider inserting
+optimization_barrier in between to prevent constant folding, but this requires
+upstream frameworks to embed optimization_barrier and downstreams to pattern
+match additional optimization_barrier. It is hard to guarantee these
+requirements on use-cases across various frameworks and hardwares.
+
+## Examples
+
+Here are examples of hybrid quantized convolution and dot_general. Hybrid
+quantized ops will get float input(or lhs) and quantized weight(or rhs) as
+operands and output float results.
+
+```mlir
+%conv = "stablehlo.convolution"(%input_f, %weight_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+```mlir
+%dot = "stablehlo.dot_general"(%lhs_f, %rhs_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+## Proposed spec changes
+
+We propose to modify operation semantics and a few constraints on quantized
+convolution and dot_general to represent weight-only quantization using hybrid
+op.
+
+### dequantize_op semantics
+
+We propose to define `dequantize_op` semantics as part of quantization
```

Comment:
Even though I understand that the name `dequantize_op` is derived from removing the quantize part from `dequantize_op_quantize`, it is a little confusing.  I feel it will be better to call this `hybrid_dequantize_then_op` and we maybe have to rename `dequantize_op_quantize` into `dequantize_then_op_then_quantize` (even though this looks verbose).  Maybe someone else can come up with a better name.

Also, I think the implementation below can be generalize for multiple implements:

```python
def hybrid_dequantize_then_op(op, *inputs):
  float_inputs = map(lambda input: input if is_float(element_type(type(input))) else dequantize(input) , inputs)
  return op(*float_inputs)
```

Reply:
+1. Another suggestion:

- `call_dequantize_op_quantize`
- `call_dequantize_hybrid_op`

makes it a little clearer IMO that all terms in the function names refer to actual function calls and not just concepts. The `_then_` alternatives are also acceptable IMO, but agreed that `dequantize_op` could be misleading.

Reply:
Both suggestions look good to me and I have no strong preference between them. But as this involves changes to existings op semantics, I would like to get opinion from StableHLO team.

@GleasonK Could you provide some input on this?

Reply:
Renamed `dequantize_op` to `hybrid_dequantize_then_op` and updated the implementation for multiple inputs.
I think `dequantize_op_quantize` is okay as is because it is not too ambiguous like `dequantize_op`.